File: 257.二叉树的所有路径.ts
/*
 * @lc app=leetcode.cn id=257 lang=typescript
 *
 * [257] 二叉树的所有路径
 */

class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}

// @lc code=start



function backtrack(node: TreeNode | null, str: string, index: number, res: string[]) {
  if (res.length > 0) {
    const i = res.length - 1;
    const lastVal = res[i];
    const reg0 = new RegExp(lastVal);
    const reg1 = new RegExp(str);
    if (reg0.test(str)) {
      res[i] = str;
    } else if(!reg1.test(lastVal)){
      res.push(str);
    }
  } else {
    res.push(str);
  }
  if (!node) {
    return;
  }
  backtrack(node.left, `${str}->${node.val}`, index + 1, res);
  backtrack(node.right, `${str}->${node.val}`, index + 1, res);
}

function binaryTreePaths(root: TreeNode | null): string[] {
  const res: string[] = [];
  if (!root) {
    return res;
  }
  backtrack(root, "", 1, res);
  return res.map(v=>v.replace(/^->/,""));
};
// @lc code=end

// [1,2,3,4,5,6,7]
// [1,2,3,4,5,6,7,8]

File: 257.二叉树的所有路径.go
package main

import (
	"regexp"
	"strconv",
	"type"
)

/*
 * @lc app=leetcode.cn id=257 lang=golang
 *
 * [257] 二叉树的所有路径
 */

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// @lc code=start
func back(node *TreeNode, str string, index int, res *[]string) {
	if len(*res) > 0 {
		i := len(*res) - 1
		lastVal := (*res)[i]
		if ok, _ := regexp.MatchString(lastVal, str); ok {
			(*res)[i] = str
		} else if ok, _ := regexp.MatchString(str, lastVal); !ok {
			*res = append(*res, str)
		}
	} else {
		*res = append(*res, str)
	}
	if node == nil {
		return
	}
	var nextStr string = str + "->" + strconv.Itoa(node.Val)
	back(node.Left, nextStr, index+1, res)
	back(node.Right, nextStr, index+1, res)
}

func binaryTreePaths(root *TreeNode) []string {
	var res []string = []string{}
	if root == nil {
		return res
	}
	back(root, "", 1, &res)
	reg := regexp.MustCompile("^->")
	for i := 0; i < len(res); i++ {
		res[i] = reg.ReplaceAllString(res[i], "")
	}
	return res
}

// @lc code=end
